### PR TITLE
implement custom metrics per DataObject (#529)

### PR DESCRIPTION
**Pull request by zzeekk**
See #529 

This is a larger refactoring of how metrics are handled.
Changes include:
- subfeeds are enriched with metrics
- spark metrics are collected synchronously
- specific metrics for jdbc, delta, iceberg and snowflake
- standardize naming for inserted/updated/deleted rows
- add metrics to SubFeedsExpressionData so it can be used in executionCondition
- ActionPipelineContext.phase is now implemented immutable
- ActionPipelineContext.currentAction added

### CODE REVIEW

File: sdl-core/src/main/scala/io/smartdatalake/metrics/SparkStageMetricsListener.scala [hunk 1]
```diff
@@ -19,30 +19,38 @@
...
+  /**
+   * Keeps track of running jobs, e.g. jobs that started but did not yet end.
+   * Map key is jobId.
+   */
+  val runningJobs: mutable.Map[Int, JobInfo] = mutable.Map.empty
 
   /**
    * Stores jobID and jobDescription indexed by stage ids.
    */
   val jobInfoLookupTable: mutable.Map[Int, JobInfo] = mutable.Map.empty
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/metrics/SparkStageMetricsListener.scala]:
```suggestion
  val stageIdToJobInfo: mutable.Map[Int, JobInfo] = mutable.Map.empty
```
A name like that makes it clear that it's just a map. I found the word "Table" confusing, I needed to open IntelliJ to understand what it does

File: sdl-core/src/main/scala/io/smartdatalake/metrics/SparkStageMetricsListener.scala [hunk 2]
```diff
@@ -84,13 +104,56 @@ private[smartdatalake] class SparkStageMetricsListener(action: Action)(implicit
...
+    metrics
+  } finally {
+    context.sparkSession.sparkContext.removeSparkListener(this)
+  }
+
+  /**
+   * Wait for Spark metrics and return main metrics from latest job.
+   */
+  def waitForMetrics(timeoutSec: Int = 10): MetricsMap = {
+    waitForSparkMetrics(timeoutSec).sortBy(_.jobInfo.id).lastOption.map(_.getMainInfos).getOrElse(Map())
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/metrics/SparkStageMetricsListener.scala:140]: what about the metrics for the other jobs that are not the last one?
Are we only interested in the metrics of the last one?
Then why are the others even stored if we are just interested in the last one?

File: sdl-core/src/main/scala/io/smartdatalake/metrics/SparkStageMetricsListener.scala [hunk 3]
```diff
@@ -84,13 +104,56 @@ private[smartdatalake] class SparkStageMetricsListener(action: Action)(implicit
...
     }
+    metrics
+  } finally {
+    context.sparkSession.sparkContext.removeSparkListener(this)
+  }
+
+  /**
+   * Wait for Spark metrics and return main metrics from latest job.
+   */
+  def waitForMetrics(timeoutSec: Int = 10): MetricsMap = {
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/metrics/SparkStageMetricsListener.scala]:
```suggestion
  def waitForMetricsOfLastJob(timeoutSec: Int = 10): MetricsMap = {
```

File: sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRun.scala [hunk 4]
```diff
@@ -112,34 +108,36 @@ private[smartdatalake] case class ActionDAGRun(dag: DAG[Action], executionId: SD
...
+        node.prepare(actionContext)
         node.outputs.map(outputDO => DummyDAGResult(outputDO.id.id))
       case x => throw new IllegalStateException(s"Unmatched case $x")
     }
   }
 
-  def init(implicit context: ActionPipelineContext): Seq[SubFeed] = {
-    context.phase = ExecutionPhase.Init
+  def init(context: ActionPipelineContext): Seq[SubFeed] = {
+    implicit val phaseContext: ActionPipelineContext = context.copy(phase = ExecutionPhase.Init)
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRun.scala:126]: how can it happen that context.phase is not Init in this situation? seems strange

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataframe/GenericDataFrame.scala [hunk 5]
```diff
@@ -231,7 +231,7 @@ trait GenericSchema extends GenericTypedObject {
     SchemaConverter.convert(this, toSubFeedType)
   }
   def equalsSchema(schema: GenericSchema): Boolean = {
-    diffSchema(schema).nonEmpty || schema.diffSchema(this).nonEmpty
+    diffSchema(schema).isEmpty && schema.diffSchema(this).isEmpty
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataframe/GenericDataFrame.scala:234]: looks like a massive bugfix! How come it seemded to have no big impact until now?

**Review by Geheiner — COMMENTED** [review]
Let's discuss this together.
I'm not sure about the impact of the phase now being immutable and on the syncronous waiting for metrics

File: sdl-core/src/main/scala/io/smartdatalake/metrics/SparkStageMetricsListener.scala (hunk 2 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/metrics/SparkStageMetricsListener.scala:140]: For debugging and maybe future detailled Spark job statistics it is interesting to keep all metrics.

File: sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRun.scala (hunk 4 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRun.scala:126]: Well, this is the beginning of the Init-Phase and we have to initialize the context correctly.

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataframe/GenericDataFrame.scala (hunk 5 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/dataframe/GenericDataFrame.scala:234]: You're right.... for now this method is only used in IcebergTableDataObject. I did some tests and found this bug.

**Comment by zzeekk** [comment]
> Let's discuss this together. I'm not sure about the impact of the phase now being immutable and on the syncronous waiting for metrics

Well, the flow gets easier like that. I discovered this possibility when implementing expectations, and think we should apply it here as well, as it also allows to better control the metrics created per DataObject, e.g. every DataObject can decide it's own metrics.

File: sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRun.scala (hunk 4 above)
Comment (Geheiner) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRun.scala:126]: And why not directly change the original context variable when you call the respective function ?, eg
actionDAGRun.init(context.copy(phase = ExecutionPhase.Init))
Or do we have it in too many places?

**Review by Geheiner — APPROVED** [review]
It looks good to me

File: sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRun.scala (hunk 4 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRun.scala:126]: Yes, it's called in many places. Like this you don't have to remember to pass a context with phase = ExecutionPhase.Init to this function.

**Pull request merged**